src/ILVerify/src/ILImporter.Verify.cs

```diff
@@ -365,17 +365,28 @@ void CheckIsValidBranchTarget(BasicBlock src, BasicBlock target)
             if (src.TryIndex != target.TryIndex)
             {
                 if (src.TryIndex == null)
-                    VerificationError(VerifierError.BranchIntoTry);
+                {
+                    // Branching to first instruction of try-block is valid
+                    if (target.StartOffset != _exceptionRegions[(int)target.TryIndex].ILRegion.TryOffset || !IsDirectChildRegion(src, target))
+                        VerificationError(VerifierError.BranchIntoTry);
+                }
                 else if (target.TryIndex == null)
                     VerificationError(VerifierError.BranchOutOfTry);
                 else
                 {
-                    if (_exceptionRegions[(int)src.TryIndex].ILRegion.TryOffset < _exceptionRegions[(int)target.TryIndex].ILRegion.TryOffset)
-                        VerificationError(VerifierError.BranchIntoTry);
+                    var srcRegion = _exceptionRegions[(int)src.TryIndex].ILRegion;
```

Comment:
Nit: ILExceptionRegion structure is relatively large (7 integers). It would be a bit more efficient to cache it in a stack local as `ref ILExceptionRegion srcRegion = ...`, or as the enclosing `ExceptionRegion srcRegion = ...` to avoid copies of the whole structure . This issue exists in the existing code as well.

Reply:
Very true. I will fix this in my next pull request.